### PR TITLE
Send a simple customer data payload

### PR DESCRIPTION
**Pull request by benjaminjb**
**Checklist:**
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
The upgrade check just sends PGO version as a header.

**What is the new behavior (if this is a feature change)?**
This adds some material to the upgrade check header, including time, number of clusters managed, and an anonymous ID.

**Other information**:
Issue [sc-12903]

### CODE REVIEW

**Comment by brandur** [comment]
@benjaminjb Hey! Just wanted to let you know that the server-side is more or less ready for consumption here.

Here are the endpoints you can hit:

* Production: https://operator-maestro.crunchydata.com/pgo-versions
* Staging: https://operator-maestro-staging.herokuapp.com/pgo-versions

I'd recommend checking the response structure of the list with `curl -i https://operator-maestro.crunchydata.com/pgo-versions` and adapting your code to it. For now, the response is extremely simple, containing only a version tag.

I also have a couple minor asks for some tweaks to the format here:

* Can you rename the header from `X-Client-Upgrade-State` to `X-Crunchy-Client-Metadata`? I think the naming is a little more apt given the payload and it's a good idea to include a vendor name in there.
* In the header payload, could we rename `id` to `deployment_id`? It's better to be a little more specific about what type of identifier this is in case we want to add other IDs in the future.

File: internal/upgradecheck/header.go
Comment (andrewlecuyer) [inline internal/upgradecheck/header.go:139]: :wrench: One issue I'm seeing here is that if the `deployment_id` is deleted from the ConifgMap (but the ConfigMap itself still exists), we'll never update the ConfigMap with what we have in memory.  Can we use a Server-Side Apply in this function instead of using `Create()`?

I do realize our `apply()` function as used for SSA is in the `postgrescluster` package.  Maybe we move it elsewhere so that it can also be used here as well?

I also wouldn't be against just creating `apply()` and `patch()` funcitons in this package that serve the specific use case of the upgrade checker (e.g. this thing only deals with ConfigMap's, so no need to worry about the workarounds for StatefulSets, Jobs and Services in the current `apply()` function).

The benefit of a SSA is that we'll always being applying our intent, which means we'll always set the `deployment_id` if its missing.  And if we need to add additional fields to the ConfigMap in the future, we can simply update our intent, and those changes will be patched (and overall this aligns with our use of SSA for creating/patching resources throughout other parts of PGO v5).

File: go.mod
Comment (benjaminjb) [reply go.mod:9]: No, it's a relatively easy swap out -- actually since apimachinery's uuid is mostly focused on types.UID (which is itself a string type), using that and just making sure everything is a string makes some things cleaner. (I didn't see us use any particular UUID package, though I feel like this has come up before...

File: internal/upgradecheck/header.go
Comment (benjaminjb) [reply internal/upgradecheck/header.go:60]: So this was an interesting case: I decided it made sense to NOT stop on error but just use a default/nil value; but I also wanted to get the errors from each individual func in order to log an apt error message; so I figured I could use individual errs for each func (as I did here) or to set the err back to nil after each time it was set to a non-nil value. 

Do you think that second option would be cleaner?

Or stepping back: what do you think about not stopping?

File: internal/upgradecheck/header.go
Comment (benjaminjb) [reply internal/upgradecheck/header.go:125]: Yes, PGO creates only one configmap, and never deletes OR updates it. (i.e., if it's there and has a value, use that value, don't overwrite that value ever).

File: internal/upgradecheck/header.go
Comment (benjaminjb) [reply internal/upgradecheck/header.go:133]: My thinking on this was that `ensureDeploymentID` always returns _some_ ID -- the worst case scenario would be that it just creates an ID and saves that in memory -- so the error becomes a message for the calling function. 

I figured it made sense to leave it up to the caller to decide "I cannot trust this ID because of the error" or shrug and just use the ID. (Which is how `generateHeader` uses `ensureDeploymentID` -- log the error and use whatever ID it gives.)

(That said, I don't know who would ever use this `ensureDeploymentID` func, and it would need to be changed a bit since the configmap name is hard-coded in; so there's definitely an argument to be made that I should just log the error here and return only the ID.)

File: internal/upgradecheck/header.go
Comment (benjaminjb) [reply internal/upgradecheck/header.go:139]: That makes sense to me, with the caveat that we don't want to overwrite our deployment_id unless it's... not a valid UUID? Like:
CM doesn't exist -- create
CM exists, has deployment id UUID -- leave
CM exists, deployment id doesn't exist or is malformed -- update

Comment (benjaminjb) [reply internal/upgradecheck/header.go:139]: And for checking that some string is a valid UUID, I'd probably use that google/uuid package

File: Makefile
Comment (benjaminjb) [reply Makefile:110]: Oh, I'll just add `createnamespaces` to these targets (deploy-dev, check-envtest-existing -- the two targets where the absence of a postgres-operator namespace keep tripping me up)

File: internal/upgradecheck/header.go
Comment (andrewlecuyer) [reply internal/upgradecheck/header.go:60]: After thinking about this a bit, I am wondering if stopping is the best approach here.  We already have retry logic built in, correct?  Or does that only retry if we can't successfully contact the version endpoint?

It does also seem as though stopping is the best way to ensure we get clean data (e.g. consistent ID's, accurate cluster counts, etc.).

Lets take a few mins to discuss following the standup this morning.

File: internal/upgradecheck/header.go
Comment (andrewlecuyer) [reply internal/upgradecheck/header.go:139]: Ahh yes!  I do see we would need the Google package to check the UUID.

Also, I concur with the following :+1: 

> CM doesn't exist -- create
> CM exists, has deployment id UUID -- leave
> CM exists, deployment id doesn't exist or is malformed -- update

File: internal/upgradecheck/header.go
Comment (andrewlecuyer) [reply internal/upgradecheck/header.go:133]: Understood - lets hash out a bit following standup, considering the thread above regarding stopping on errors.

As it stands, I personally like the idea of logging a warning/info message and not returning an error if a valid deployment ID is returned. In other words, if the function is returning some valid UUID we can/will use, then I feel as though there shouldn't be an error (I'm also not sure how the consumer would be able to tell if the UUID should be used simply based on the error message ... e.g. if I get a UUID but also some transient Kube error, what does that mean?  I.e. can I use that to determine if its a valid UUID?).

File: internal/config/config.go
Comment (benjaminjb) [reply internal/config/config.go]: Updating this according to our chat yesterday and updated all testing make targets

File: internal/upgradecheck/header.go
Comment (benjaminjb) [reply internal/upgradecheck/header.go]: Following discussion, we decided to stick with a standard Go pattern of logging the error rather than returning an error that is then just logged by the calling function.

File: cmd/postgres-operator/main.go
Comment (andrewlecuyer) [inline cmd/postgres-operator/main.go:92]: :question: I have noticed that if there an attempt to use the client before the manager has started (i.e. before `mgr.Start()` is called a few lines down), and error will be thrown because the cache's for the manager/client have not yet synced.

Have you run into anything like this with your testing?  It seems like this would only _potentially_ be a concern for the first attempt to hit the endpoint when PGO starts.

Comment (benjaminjb) [reply cmd/postgres-operator/main.go:92]: I have not seen that sort of error in testing this (either running the tests or running the program), but that might be just luck so far.

Actually, I just put in a sleep before the `mgr.Start` to try to force an error and that "successfully" errored; so I put in a WaitForCacheSync for-loop and that solved the error.

(I'm also beginning to think I should just pass the mgr in to retrieve the context, client, and cache, but that kind of obscures what we need, so I'll leave it for now, with some 🌱 for the future.)

File: internal/upgradecheck/header.go
Comment (andrewlecuyer) [inline internal/upgradecheck/header.go:120]: :wrench: Since this function is now using SSA, I would expect the intent to be applied each time this logic is run (instead of returning the current ConfigMap as is done here).

Therefore, in the `default` case here where the ConfigMap exists and has valid ID, I expect that we would simply proceed on to the logic below to generate the intent, specifically using the existing ID we parse here as `deployment_id` in the intent.  That means we will always be applying our full intent each time this is run (which most of the time will not change, but I could see us eventually adding other labels, annotation and/or data to the intent).

File: internal/naming/names.go
Comment (andrewlecuyer) [inline internal/naming/names.go:487]: :wrench: Can we drop `configmap` from the name? (feels like a bit of stutter otherwise).

:thinking: Just wondering aloud if we should name more generically - e.g. just `pgo` - as the "PGO ConfigMap" (which maybe could eventually be used for other things - though nothing else comes to mind at the moment (which is why I def don't have a problem sticking with the current naming).

File: cmd/postgres-operator/main.go
Comment (andrewlecuyer) [reply cmd/postgres-operator/main.go]: Sounds great, definitely on board with using `WaitForCacheSync()` here!

**Review by andrewlecuyer — CHANGES_REQUESTED** [review]

File: internal/upgradecheck/http.go
Comment (andrewlecuyer) [inline internal/upgradecheck/http.go:163]: :question: Is the `for` loop needed here?  From what I can tell `WaitForCacheSync()` will simply block and not return until the cache finally syncs, but just let me know if there is anything I'm missing.

Comment (benjaminjb) [reply internal/upgradecheck/http.go]: As far as I can tell, yeah, `WaitForCacheSync` is well named -- I can remove the loop, it makes it cleaner without, I just put it in because it felt a little odd to run a function without checking the return.

Actually, the more I look at this without any for loop, I wonder about the off chance that syncing will encounter an error: if we're in a loop, then we just keep erroring until we don't. I almost feel like this would be another good place for a retry with an exponential backoff. (The other solution is to error and exit the process; an operator in that condition would never check for an upgrade.)

My feeling right now is... maybe remove the for-loop, and create a ticket for further enhancements?

Comment (andrewlecuyer) [reply internal/upgradecheck/http.go]: My thought is that if `WaitForCacheSync()` returns `false`, we would simply log an error and then _not_ proceed with running the "check for upgrade" logic.

My thinking here is that the `manager` will also fail to start in the event the caches don't sync successfully, so we're probably looking at a failure to start PGO as a whole anyway (and therefore likely a container restart).

File: internal/upgradecheck/http_test.go
Comment (andrewlecuyer) [inline internal/upgradecheck/http_test.go:170]: :thinking: Can we use `envtest` here instead of `fake`?  It is specifically the `CacheWithWait` interface that I'm thinking about, i.e. if we use `envtest` and start the manager, can we just then grab/use the client as we do in our other tests?  With this approach the cache should always sync before we grab the client for use in our tests, so wondering if that would allow us to avoid the need to create the interface simply to facilitate mocking.

Maybe this is something we can consider for a future commit (switch to `envtest`)?  I definitely don't want to blow up the current tests at this point or anything (since there are a lot of great tests in here), so lets be sure to discuss before making any changes.  Also, if anything that I'm missing, just let me know.

**Review by andrewlecuyer — APPROVED** [review]
I have one comment about using `envtest` that I would like to hash out before merging, so just let me know when you have a chance to discuss

File: internal/upgradecheck/http_test.go
Comment (benjaminjb) [reply internal/upgradecheck/http_test.go:170]: I made a note to change this in the future; I did a quick version of changing it over and it was mostly clean but there were some failures, so as we discussed, I'll merge in now and we can iterate at a later point.

**Pull request merged**
